### PR TITLE
core: Increase precision of server group resolution (when we know the…

### PR DESCRIPTION
**Pull request by duftler**
… exact location type, use that information).

### CODE REVIEW

**Comment by duftler** [comment]
@ttomsu please review.
@cfieber @anotherchrisberry fyi

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy [hunk 1]
```diff
@@ -79,7 +79,15 @@ class TargetServerGroup {
       }
      }
 
-    static Location locationFromServerGroup(Map<String, Object> serverGroup) {
+    static Location locationFromServerGroup(Map<String, Object> serverGroup, Location.Type exactLocationType) {
+      if (exactLocationType == Location.Type.ZONE) {
```
Comment (ttomsu) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: A switch() would look nicer here.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupResolver.groovy [hunk 2]
```diff
@@ -112,10 +112,18 @@ class TargetServerGroupResolver {
...
 
-    def tsg = tsgs.find {
-      def location = it.getLocation().value
-      return (stage.context.region == location) || (stage.context.zone == location) || (stage.context.namespace == location)
-    }
+    // TODO(duftler): Confirm that every stage.context will have a targetLocation at this point, and then drop the else branch.
+    def targetLocation = stage.context.targetLocation
+    def tsg =
+      targetLocation
+      ? tsgs.find {
```
Comment (ttomsu) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupResolver.groovy]: nit: ternary operator on the previous line, same with the colon below.

**Comment by ttomsu** [comment]
LGTM with comments

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy (hunk 1 above)
Comment (duftler) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroup.groovy]: Done.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupResolver.groovy (hunk 2 above)
Comment (duftler) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupResolver.groovy]: bah

**Pull request merged**